### ISSUE TITLE
Any plans to re-enable "public" usage, and is the OSS version meant for use?

### ISSUE THREAD

**Issue by mipearson** [sourcegraph/sourcegraph#6783]
I was looking at Sourcegraph, specifically the open source version, as a way to provide code search beyond what Github offers across all of our organisation's 500+ private repositories.

Per-user features were not needed, nor was security: we'd put it behind our own firewall that we use to lock off internal resources from the wider internet.

The decision to remove the "public" option, along with the lack instructions on how to install the OSS version, seem commercially motivated: forcing people to sign up as a user means the user can be counted, and using the image available from Dockerhub prompts the user to leverage Enterprise-only features.

That's entirely understandable - but it feels like the OSS "plan" isn't one at all - it's meant only for customers wishing to add features rather than self-host Sourcegraph.

Is there a way to disable user authentication other than locking ourselves to version 3.7.2?

Is the Sourcegraph OSS version intended for real use rather than customers contributing features or fixes back to Sourcegraph?

**Comment by slimsag** [comment]
I work here, but am posting this on my free time because it's an issue I care about. What I say here is my unbiased and honest opinions (I am sure I sound like a fake Amazon review at this point, though).

<img src="https://user-images.githubusercontent.com/3173176/69493921-75954980-0e72-11ea-9e7b-a34fd9a07180.png" height=200px></img>

> Is there a way to disable user authentication other than locking ourselves to version 3.7.2?

For now, this is the only way. We'll find a way to support this use case (but I can't give a concrete timeline).

We didn't remove this out of malice or any profiteering-driven motives, though. We originally had this option just to support sourcegraph.com, but some customers enabled it (thinking "I don't have to configure auth, great") and didn't realize it in fact disabled many enterprise features (like precise code intel, saved searches, the git blame toggle and various extensions that require per-user configuration). You can even see [in the original PR](https://github.com/sourcegraph/sourcegraph/pull/5507#issuecomment-529631263) where we concluded the only thing we cared about here was that the user experience was good, not that we were trying to extract some $$$ out of people or something.

You can also see in the final PR where we were considering removing it that I suggested there might be [some interesting OSS use cases](https://github.com/sourcegraph/sourcegraph/pull/5514#issuecomment-529650218) we would miss out on like a truly-public sourcegraph instance, but I and my coworkers just didn't think of the use case of a private-authless instance in this final PR. I'll elaborate on why we didn't think of this below

> Is the Sourcegraph OSS version intended for real use rather than customers contributing features or fixes back to Sourcegraph?

1. We have not historically had many (any?) pure-OSS users, so it is rough around the edges no doubt. This isn't because we want it to be, just because time is finite and we have to tailor what we work on to the 99% of users obviously. It should work, though, and we want it to work.
2. We have conceded to not publish the pure-OSS images on Docker Hub, though. This is similar to what VS Code, Chrome, etc. do and is driven by a few factors, including:
    - Our ability to include enterprise features and unlock them easily in the app upon license upgrades (we could do this in an OSS image technically, but we wouldn't want an "OSS" sourcegraph image to include a bunch of mentions about enterprise features and buying licenses etc.
    - To avoid confusion about what the best starter version of sourcegraph is for enterprise users ("I am running the OSS one, how do I enable X?")

### TL;DR:

- No, the OSS version is not intended to just be 'for customers contributing features or fixes back'.
- Yes, we want the OSS version to be easily built from source and usable in real production environments.
- Yes, there are some rough edges (lack of proper docs / build steps mostly #6789 #6790 ) that we should work out soon. This is just because we haven't had many pure-OSS users before and have neglected those build scripts a bit.

### This issue

On the topic of getting this option back into the product; I propose the following:

We bring back `auth.public` but under a different name and with clear warning documentation about its limitations:

> auth.degraded.public
>
> Allows anonymous visitors full read access to repositories, code files, search, and other data (except site configuration).
>
> IMPORTANT: Because many core features rely on persisted user settings, anonymous usage leads to a degraded experience for most users. As a result, for self-hosted private instances it is preferable for all users to have accounts.
>
> SECURITY WARNING: If you enable this, you must ensure that only authorized users can access the server (using firewall rules or an external proxy, for example).\n\nRequires usage of the builtin authentication provider.

**Comment by mipearson** [comment]
Thank you so much for your detailed response! :)

I want to point out that I personally don't feel that there's anything wrong with the idea that the OSS variant is optimised towards customers wanting to contribute their own features or fixes, nor with a "free" version being designed to encourage people to license the paid product. My worry was that this wasn't properly signposted.

Fortunately, that was my misunderstanding, and I'm looking forward to giving this a proper try on version 3.7.2 to see if it fits our organisation's needs for searching across our repositories.

**Comment by aeb-sia** [comment]
We managed to get an open source docker image of sourcegraph and got it up and running.

I connected sourcegraph to a private gitlab instance and as an admin user I can search through all repositories. We tried adding users only to find out we can't allow them to search the repositories. We don't want any authorization. All users should be allowed to access all repositories. Ideally we would just allow unauthenticated access, but we can live with creating users/allowing signups.

Is the only way to allow all users access to all repositories by promoting them to admins?

**Comment by simhnna** [comment]
BTW here is an open source image that I'm building and using https://hub.docker.com/r/simhnna/sourcegraph-oss

**Comment by bhush9** [comment]
@slimsag Any progress on this issue? We at [KDE](https://www.kde.org) are looking into deploying opensource version and would be really nice to have support for this.

**Comment by sqs** [comment]
I started a branch at https://github.com/sourcegraph/sourcegraph/pull/11575 to expose an option to make Sourcegraph instances publicly accessible (like [https://sourcegraph.com](https://sourcegraph.com/search).

@bhush9 Do you know what else KDE would need to run a public Sourcegraph instance beyond this?

**Comment by crscillitoe** [comment]
Any update on this? I tried hard locking my version to 3.7.2, but I receive "(root): Additional property auth.public is not allowed" when I try to set `auth.public` to true.

**Comment by crscillitoe** [comment]
Going to bump asking for response here, as this still hasn't been addressed, I'm really looking for a solution to be able to use this flag on my personal deployment for my OSS

**Comment by slimsag** [comment]
@christinaforney @sqs let's chat about how we can fit the remaining work here into our roadmap? or maybe we can put together an RFC/spec with what work is needed here?

**Comment by hitzhangjie** [comment]
> BTW here is an open source image that I'm building and using https://hub.docker.com/r/simhnna/sourcegraph-oss

Hi, How do you make the sourcegraph accessible for all users without registration.

Thanks in advance.

**Comment by aexvir** [comment]
@slimsag could there be some update on this issue?

**Comment by jensim** [comment]
I was drawn here by the question about OSS docker image..
I made my own release pipe for the OSS image, and it has mostly worked great :-) But the drawback of "owning" code that is 99.99% dependent on another player that does constant updates, is a recipe for just owning uncomfortable problems. I would love it if something like this was available from the official pipeline.
https://github.com/jensim/sourcegraph-release-train

**Comment by mrvaghani** [comment]
Hi there,

Like many others on this thread, I ended up here due to not having enough documentation to successfully build/deploy the OSS version of sourcegraph. Reviewing some other open issues, it seems that sourcegraph is aware of the lack and already has a few things in their backlog. 

Instead of having many different open issues discussing the same thing, may I suggest selecting a single issue as the main thread to post updates and keep track of the progress?

Here are some of the issues(open/closed) that I found related to this topic:

1. [Documentation for building OSS and enterprise images is missing](https://github.com/sourcegraph/sourcegraph/issues/6790)
2. [Our public derivative Docker images should be OSS](https://github.com/sourcegraph/sourcegraph/issues/2299)
3. [Open Source Plan](https://github.com/sourcegraph/sourcegraph/issues/6709)
4. [Building enterprise and OSS images outside of CI is broken](https://github.com/sourcegraph/sourcegraph/issues/6789)
5. [Any plans to re-enable "public" usage, and is the OSS version meant for use?](https://github.com/sourcegraph/sourcegraph/issues/6783)

I am willing to help test documentation for OSS and potentially also assist with having OSS image on sourcegraph's dockerhub repo. Please let me know if there is anything I can do to assist with speeding up the OSS documentation.

**Comment by jensim** [comment]
@mrvaghani try my release train? https://github.com/jensim/sourcegraph-release-train

**Comment by mrvaghani** [comment]
@jensim I did look at your release train and your public docker image and I was able to get it working without any issues. It's nicely documented and very easy to follow.

My only concern is that your build uses files in enterprise folder, along with some test license files that I am not sure are meant for opensource use. Sourcegraph specifically mentions in their [license section](https://github.com/sourcegraph/sourcegraph#license) that everything under enterprise folder is not part of OSS. I was kinda hoping for an official OSS docker image from sourcegraph that we can use in the enterprise environment.

I have created an [issue](https://github.com/jensim/sourcegraph-release-train/issues/1) in your repo regarding this if you want to take the discussion there.

**Comment by jensim** [comment]
@mrvaghani thats just a little experiment that I ran as a little pen test.. The docker release is not using any enterprise parts.

**Comment by ltagliamonte** [comment]
Hello Folks,
i've built myself a v4.2.1 with no enterprise and i'm trying to add [oauth2-proxy](https://oauth2-proxy.github.io/) obviously it is not working because i can't disable auth.
Anyone here with a workaround?

**Comment by simhnna** [comment]
> Hello Folks,
> i've built myself a v4.2.1 with no enterprise and i'm trying to add [oauth2-proxy](https://oauth2-proxy.github.io/) obviously it is not working because i can't disable auth.
> Anyone here with a workaround?
> 

I've developed a couple of patches for the OSS version that patched away broken things in the UI. One of the patches is for authenticating users via http headers. Nothing fancy, but you should be able to roll your own based on that code... https://github.com/simhnna/sourcegraph-release-train/blob/main/patches/auth.patch

**Comment by ltagliamonte** [comment]
@simhnna thanks for the link!
Is there any configuration i need to set to enable the http headers and the login as admin?

**Comment by simhnna** [comment]
There's no configuration options at all. The used headers are hardcoded and it's always on. Could be improved, but I wanted to have a minimal patch

**Comment by jdorfman** [comment]
Hey all! This is now fixed via https://github.com/sourcegraph/sourcegraph/pull/52440 / @erzhtor

I'm going to close this issue. Feel free to leave [product feedback here](https://github.com/sourcegraph/sourcegraph/discussions/new/choose).

https://github.com/sourcegraph/sourcegraph/assets/6717049/344f9b1e-f54a-4148-9bc8-f0f3d3bb4cd7

**Comment by nicolas17** [comment]
What is a license tag? How do I enable public / unauthenticated access in the *open source* version?
